flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksNativeFullSnapshotStrategy.java

```diff
@@ -0,0 +1,285 @@
+    @Nonnull private final UUID backendUID;
+
+    /** The help class used to upload state files. */
+    private final RocksDBStateUploader stateUploader;
+
+    public RocksNativeFullSnapshotStrategy(
+            @Nonnull RocksDB db,
+            @Nonnull ResourceGuard rocksDBResourceGuard,
+            @Nonnull TypeSerializer<K> keySerializer,
+            @Nonnull LinkedHashMap<String, RocksDbKvStateInfo> kvStateInformation,
+            @Nonnull KeyGroupRange keyGroupRange,
+            @Nonnegative int keyGroupPrefixBytes,
+            @Nonnull LocalRecoveryConfig localRecoveryConfig,
+            @Nonnull File instanceBasePath,
+            @Nonnull UUID backendUID,
+            @Nonnull RocksDBStateUploader rocksDBStateUploader) {
+        super(
+                DESCRIPTION,
+                db,
+                rocksDBResourceGuard,
+                keySerializer,
+                kvStateInformation,
+                keyGroupRange,
+                keyGroupPrefixBytes,
+                localRecoveryConfig,
+                instanceBasePath,
+                backendUID.toString().replaceAll("[\\-]", ""));
+        this.backendUID = backendUID;
+        this.stateUploader = rocksDBStateUploader;
+    }
+
+    @Override
+    public SnapshotResultSupplier<KeyedStateHandle> asyncSnapshot(
+            NativeRocksDBSnapshotResources snapshotResources,
+            long checkpointId,
+            long timestamp,
+            @Nonnull CheckpointStreamFactory checkpointStreamFactory,
+            @Nonnull CheckpointOptions checkpointOptions) {
+
+        if (snapshotResources.stateMetaInfoSnapshots.isEmpty()) {
+            return registry -> SnapshotResult.empty();
+        }
+
+        return new RocksDBNativeFullSnapshotOperation(
+                checkpointId,
+                checkpointStreamFactory,
+                snapshotResources.snapshotDirectory,
+                snapshotResources.stateMetaInfoSnapshots);
+    }
+
+    @Override
+    public void notifyCheckpointComplete(long completedCheckpointId) {
+        // nothing to do
+    }
+
+    @Override
+    public void notifyCheckpointAborted(long abortedCheckpointId) {
+        // nothing to do
+    }
+
+    @Override
+    protected PreviousSnapshot snapshotMetaData(
+            long checkpointId, @Nonnull List<StateMetaInfoSnapshot> stateMetaInfoSnapshots) {
+        for (Map.Entry<String, RocksDbKvStateInfo> stateMetaInfoEntry :
+                kvStateInformation.entrySet()) {
+            stateMetaInfoSnapshots.add(stateMetaInfoEntry.getValue().metaInfo.snapshot());
+        }
+        return EMPTY_PREVIOUS_SNAPSHOT;
+    }
+
+    @Override
+    public void close() {
+        stateUploader.close();
+    }
+
+    /** Encapsulates the process to perform a full snapshot of a RocksDBKeyedStateBackend. */
+    private final class RocksDBNativeFullSnapshotOperation
+            implements SnapshotResultSupplier<KeyedStateHandle> {
```

Comment:
I think the code duplication still looks a bit ugly. We can introduce another abstract `RocksDBSnapshotOperation`.
Maybe we can introduce this abstract class below in `RocksDBSnapshotStrategyBase` (also let `RocksDBSnapshotStrategyBase` include `backendUID`):

~~~java
   protected abstract class RocksDBSnapshotOperation implements SnapshotResultSupplier<KeyedStateHandle> {
        /** Id for the current checkpoint. */
        protected final long checkpointId;

        @Nonnull protected final CheckpointStreamFactory checkpointStreamFactory;

        /** The state meta data. */
        @Nonnull protected final List<StateMetaInfoSnapshot> stateMetaInfoSnapshots;

        /** Local directory for the RocksDB native backup. */
        @Nonnull protected final SnapshotDirectory localBackupDirectory;

        protected RocksDBSnapshotOperation(
                long checkpointId,
                @Nonnull CheckpointStreamFactory checkpointStreamFactory,
                @Nonnull SnapshotDirectory localBackupDirectory,
                @Nonnull List<StateMetaInfoSnapshot> stateMetaInfoSnapshots) {
            this.checkpointId = checkpointId;
            this.checkpointStreamFactory = checkpointStreamFactory;
            this.stateMetaInfoSnapshots = stateMetaInfoSnapshots;
            this.localBackupDirectory = localBackupDirectory;
        }

        protected Optional<KeyedStateHandle> getLocalSnapshot(
                @Nullable StreamStateHandle localStreamStateHandle,
                Map<StateHandleID, StreamStateHandle> sharedStateHandleIDs) throws IOException {
            final DirectoryStateHandle directoryStateHandle =
                    localBackupDirectory.completeSnapshotAndGetHandle();
            if (directoryStateHandle != null && localStreamStateHandle != null) {
                return Optional.of(
                        new IncrementalLocalKeyedStateHandle(
                                backendUID,
                                checkpointId,
                                directoryStateHandle,
                                keyGroupRange,
                                localStreamStateHandle,
                                sharedStateHandleIDs));
            } else {
                return Optional.empty();
            }
        }
    }
~~~

Reply:
By doing so, we can continue to reduce the code duplication in `RocksNativeFullSnapshotStrategy` and `RocksNativeFullSnapshotStrategy`.